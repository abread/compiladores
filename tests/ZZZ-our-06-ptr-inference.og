int a = nullptr;

auto f(int i, ptr<auto> vp, int vpi) {
	ptr<real> rp = nullptr + 0x8000;
	ptr<string> sp = nullptr + 0x9000;

	if i == 0 then
		return 1, 2.1, vp, rp;
	else
		return 1.1, 2, vpi, sp;
}

public int og() {
	{
		ptr<real> a = nullptr + 7;
		ptr<auto> b = a;
		int c = b;

		writeln b;
		writeln c;
	}

	{
		ptr<real> ap = nullptr + 69;
		ptr<auto> vap = ap;

		writeln vap;
		writeln vap+1;
		writeln 1+vap;
		writeln vap-1;
		writeln 1-vap;
		writeln (ap+1)-ap;
	}

	{
		ptr<int> p = nullptr + 42;
		ptr<auto> vp = p;
		auto a, b = f(0, vp, vp), f(1, p, vp);

		writeln a@1;
		writeln a@2;
		writeln a@3;
		writeln a@4;
		writeln b@1;
		writeln b@2;
		writeln b@3;
		writeln b@4;
	}


	return 0;
}
