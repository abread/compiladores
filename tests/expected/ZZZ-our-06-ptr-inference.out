7
7
69
70
70
68
-68
8
1
2.1
42
32768
1.1
2
42
36864
